File: android/measure/src/main/java/sh/measure/android/attributes/ComputeOnceAttributeProcessor.kt
package sh.measure.android.attributes

/**
 * Generates the attributes once and then caches them. Subsequent calls to [appendAttributes] will return the
 * cached attributes. This is useful for attributes that are expensive to compute and are not
 * expected to change.
 *
 * Implementations should override [computeAttributes] to compute the attributes and do not need
 * to override [appendAttributes].
 */
internal abstract class ComputeOnceAttributeProcessor : AttributeProcessor {
    private var isComputed = false
    private lateinit var cachedAttrs: Map<String, Any?>

    override fun appendAttributes(attributes: MutableMap<String, Any?>) {
        if (!isComputed) {
            this.cachedAttrs = computeAttributes()
            attributes.putAll(this.cachedAttrs)
            isComputed = true
        }
        attributes.putAll(cachedAttrs)
    }

    /**
     * Compute the attributes and return them as a map.
     */
    abstract fun computeAttributes(): Map<String, Any?>
}

File: ios/Sources/MeasureSDK/Swift/Attribute/ComputeOnceAttributeProcessor.swift
//
//  ComputeOnceAttributeProcessor.swift
//  MeasureSDK
//
//  Created by Adwin Ross on 03/09/24.
//

import Foundation

protocol ComputeOnceAttributeProcessor {
    func computeAttributes()
    func updateAttribute(_ attribute: Attributes)
}

/// Generates the attributes once and then caches them. Subsequent calls to [appendAttributes] will return the cached attributes.
/// This is useful for attributes that are expensive to compute and are not expected to change.
///
/// Implementations should override [computeAttributes, updateAttribute] to compute the attributes and do not need to override [appendAttributes].
///
class BaseComputeOnceAttributeProcessor: AttributeProcessor, ComputeOnceAttributeProcessor {
    private var isComputed = false

    func appendAttributes(_ attribute: Attributes) {
        if !isComputed {
            computeAttributes()
            isComputed = true
        }
        updateAttribute(attribute)
    }

    func updateAttribute(_ attribute: Attributes) {
        fatalError("Subclasses must override updateAttribute()")
    }

    func computeAttributes() {
        fatalError("Subclasses must override computeAttributes()")
    }
}
